Enonic UI: Add LocalStorage sync for stores #9292

reisfmb opened the pull request.

reisfmb committed:
Enonic UI: Add LocalStorage sync for stores #9292

Implemented syncStore utility
Used syncStore utility on projects (activeProjectId) store
Used activeProject from projects store on Sidebar for automatically updates on ProjectIcon
Minor improvements and linting fixes

reisfmb commented:
@edloidas 

Suggestion of merging order for epic branch `issue-9278`:

1. #9302 
2. #9309 
3. #9325

edloidas committed:
Enonic UI: Add LocalStorage sync for stores #9292

Improved errors handling when connecting to store.
Removed `syncPath` as it introduce excessive complexity and not actually flexible.
Added support for sessionStorage (useful to sync data between isolated parts of the same tab).
Moved throttle function to different utility file.

edloidas reviewed (approved):
Had to spend some time improving the current implementation. Can be considered done, but needs to wait for previous PRs to be merged first.

edloidas committed:
Enonic UI: Add LocalStorage sync for stores #9292

Fixed issue, that storage being updated before actually initializing the store.
Updated projects store management.
Added basic config store to store project data on init.

reisfmb commented:
Here are my thoughts on the improvements you did on synching:

## 1 - Error handling
Amazing 👏 

I was not aware of that many edge cases for when local storage could fail or was not available... Could not have done better myself!

## 2 - Tab synching
Didn't think of that... nice addition as well!

## 3 - Partial store synching
After some thought, in my opinion, the best approach here would be the following:

We de not export `syncStore`, but rather export wrapper functions that will manually set encode / decode functions based on the nanostore type (Atom, Map, DeepMap).

### Atom
For Atom, I think the encoder / decoder should be the default ones:

```ts
encode = (data: V): string => JSON.stringify(data),
decode = (raw: string): V => JSON.parse(raw) as V,
```

### Map
For Map, you've already done like what I had in mind.

### DeepMap
For DeepMap, my suggestion would be to have the decode as in the `syncMapStore`, but the encode like so:
```ts
 const encode = isPartialSync
        ? (data: M): string => {
              let partial: Partial<M> = {};
              
              for (const key of keys) {
                    const pathValue = getPath(data, key);
                    partial = setPath(partial, key, pathValue)
              }
              
              return JSON.stringify(partial);
          }
        : (data: M): string => JSON.stringify(data);
```

Where `getPath` and `setPath` are nanostore functions.

In this specific case, the `options.keys` would be of type: `AllPaths<M>[]`, which provides proper typings for all paths in the DeepMap store.

---

I think its important to now allow custom encoder / decoder, because if we do it like so, the data on the storage might not have the same structure as the data of the store (or a path of it)... I mean, nothing blocks me to do something like:

```ts
syncStore($myMapStore, 'my-map-store', {
    encode: (data) => JSON.stringify({"a-non-existing-key-on-$myMapStore": 123}),
});
```

In my point of view, allowing that is problematic because, it not only stores in the storage a data that differs from the store, as it can also change the data in the store (`store.set` on initialValue and tabSync).

## 4 - Store subscription

The current subscription to the store is done by `store.listen`. That is triggered on any store change... Now, if our store is Map or DeepMap, we only want to trigger our callback (`throttledWrite`) when a specific key / path changes. Nanostores have this `listenKeys` that we can use! The problem is that implementing that change is not straightforward, and would require quite some changes in code. 

## 5- Other less relevant points

- If we only sync a key of a Map store or a path of a DeepMap store, I guess it'd make sense to consider that key / path on the storage key as well, since the data in the storage is not the whole store itself, but rather a section of it.
- Remove console logs such as `console.log('removeFromStorage <${storageKey}>');`?!

reisfmb commented:
Now, my thoughts on the changes you did on `projects.store.ts`:

If my point 5 [here](https://github.com/enonic/app-contentstudio/pull/9325#issuecomment-3500600648) is not implemented, I'd rename the localStorage key:

```ts
syncMapStore($projects, 'projects.activeProjectId', {
    keys: ['activeProjectId'],
    loadInitial: true,
    syncTabs: true,
});
```

I also suggest to move this:
```ts
const projectIdFromUrl = getProjectIdFromUrl();
if (projectIdFromUrl) {
    $projects.setKey('activeProjectId', projectIdFromUrl);
}
```

To the scope of`updateActiveProject`:
```ts
function updateActiveProject(): void {
    if ($activeProject.get()) {
        return;
    }

    const {projects} = $projects.get();

    if (projects.length === 1) {
        setActiveProject(projects[0]);
        return;
    }

    const projectFromUrl = projects.find((p) => getProjectId(p) === getProjectIdFromUrl());
    if (projectFromUrl) {
        setActiveProject(projectFromUrl);
    }
}
```

Consider the case where we have > 2 projects. If we don't do that change above, upon deleting the selected one, none will be selected as the active one.

edloidas commented:
**Regarding comments:**

3. Partial store synching

> We de not export syncStore, but rather export wrapper functions that will manually set encode / decode functions based on the nanostore type (Atom, Map, DeepMap).

Not really sure how that would work.

4. Store subscription

> The current subscription to the store is done by store.listen. That is triggered on any store change

Yeah, I thought about this too. It's a balance between simplicity and full coverage of edge cases — and there are *a lot* of them. I haven't found a simple way to organize the code without adding significantly more logic to trigger updates only on specific key changes. A throttle should be more than enough to avoid performance issues (and our stores are simple enough anyway). I'd rather keep things simple. If we go deeper here, it should probably be a separate library — something we could hack on in spare time. The devs have their own [implementation](https://github.com/nanostores/persistent) of localStorage sync, but I don't find it that flexible.

5. Other less relevant points

> make sense to consider that key / path on the storage key as well

Too complex. DeepMaps are powerful, but very tricky to cover properly. I don't think we should try to use them with partial sync at all right now. They're also difficult to cover with types. So the approach is: we stick to atoms and maps primarily. DeepMaps are treated like regular maps for now.

> Remove console

Yep, just leftovers from debugging.

6.

> Consider the case where we have > 2 projects. If we don't do that change above, upon deleting the selected one, none will be selected as the active one.

Good point — we should add that as you suggested. But this shouldn't replace the `projectIdFromUrl` logic above, because the example won't run if we reload the page with a different URL when an active project is already set.

**Summarizing:**

* We drop DeepMap support for now and don't worry about key-level subscriptions. Too long to implement properly, low immediate value. Could be a separate side-project library later.
* Remove logs.
* Add your `projectIdFromUrl` change.

reisfmb committed:
Enonic UI: Add LocalStorage sync for stores #9292

Removed export from syncStore and implemented syncAtomStore

reisfmb committed:
Enonic UI: Add LocalStorage sync for stores #9292

Minor improvements on projects store

reisfmb committed:
Enonic UI: Add LocalStorage sync for stores #9292

Fixed active project name for projects without lang

edloidas committed:
Enonic UI: Add LocalStorage sync for stores #9292

edloidas commented:
Great work, Bruno
